email/email.go

```diff
@@ -0,0 +1,311 @@
+	}
+
+	smtpUsername, got = os.LookupEnv("SMTP_USERNAME")
+	if !got {
+		log.Panic("SMTP_USERNAME not set (set DISABLE_SEND_EMAIL=1 to disable)")
+	}
+
+	smtpPassword, got = os.LookupEnv("SMTP_PASSWORD")
+	if !got {
+		log.Panic("SMTP_PASSWORD not set (set DISABLE_SEND_EMAIL=1 to disable)")
+	}
+}
+
+type VerificationMetadata struct {
+	RequestUserAgent string
+	RequestIpAddress string
+	RequestTime      time.Time
+}
+
+// SendVerificationEmails iterates through the email addresses on the given key and works out
+// whether to send each one a verification email.
+// If so, it renders and sends the verification email, and records a new verification in the
+// database.
+func SendVerificationEmails(
+	txn *sql.Tx, publicKey *pgpkey.PgpKey, meta VerificationMetadata) error {
+
+	for _, email := range publicKey.Emails(true) {
+		shouldSend, err := shouldSendVerificationEmail(txn, email)
+		if err != nil {
+			return err
+		} else if shouldSend {
+			if err := sendVerificationEmail(txn, email, publicKey, meta); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func sendVerificationEmail(
+	txn *sql.Tx, emailAddress string, publicKey *pgpkey.PgpKey,
+	meta VerificationMetadata) error {
+
+	verifySecretUUID, err := datastore.CreateVerification(
+		txn, emailAddress, publicKey.Fingerprint(),
+		meta.RequestUserAgent,
+		meta.RequestIpAddress,
+		meta.RequestTime,
+	)
+	if err != nil {
+		return err
+	}
+
+	emailTemplateData := verifyEmail{
+		Email:            emailAddress,
+		VerificationUrl:  makeVerificationUrl(*verifySecretUUID),
+		RequestIpAddress: meta.RequestIpAddress,
+		RequestTime:      meta.RequestTime,
+		KeyFingerprint:   publicKey.Fingerprint().Hex(),
+		KeyCreatedDate:   publicKey.PrimaryKey.CreationTime,
+	}
+
+	email := email{
+		to:      emailAddress,
+		from:    "Fluidkeys <verify@mail.fluidkeys.com>",
+		replyTo: "Fluidkeys Security <security@fluidkeys.com>",
+		bcc:     "hello@fluidkeys.com",
+	}
+
+	if err := email.renderSubjectAndBody("verify", emailTemplateData); err != nil {
+		return fmt.Errorf("error rendering email: %v", err)
+	}
+
+	log.Printf("email.htmlBody: %s\n", email.htmlBody)
+
+	if err := email.send(); err != nil {
+		return fmt.Errorf("error sending mail: %v", err)
+	}
+	return nil
+}
+
+// shouldSendVerificationEmail returns true if an email address should receive a new verification
+// email
+func shouldSendVerificationEmail(txn *sql.Tx, email string) (bool, error) {
+	_, alreadyLinked, err := datastore.GetArmoredPublicKeyForEmail(email)
+	if err != nil {
+		return false, err
+	}
+	if alreadyLinked {
+		// 1. it's linked to the same key, in which case there's
+		//    nothing to do
+		// 2. it's linked to one key, and this is a request
+		//    to link the email to a *different* key, which we don't
+		//    currently allow. The email_key_link must be deleted
+		//    before the email can be linked again. Note that this
+		//    happens if the whole linked *key* is deleted.
```

Comment:
Might it be worth logging here, like you do in the case of `hasActiveVerification`?

Reply:
Let's try it! 1463a50a0cbfeb2483f8a3876fe3404c3b132021